wit/store.wit

```diff
@@ -0,0 +1,106 @@
+/// 
+/// Data consistency in a key value store refers to the guarantee that once a write operation
+/// completes, all subsequent read operations will return the value that was written.
+/// 
+/// Any implementation of this interface must have enough consistency to guarantee "reading your
+/// writes." In particular, this means that the client should never get a value that is older than
+/// the one it wrote, but it MAY get a newer value if one was written around the same time. These
+/// guarantees only apply to the same client (which will likely be provided by the host or an
+/// external capability of some kind). In this context a "client" is referring to the caller or
+/// guest that is consuming this interface. Once a write request is committed by a specific client,
+/// all subsequent read requests by the same client will reflect that write or any subsequent
+/// writes. Another client running in a different context may or may not immediately see the result
+/// due to the replication lag. As an example of all of this, if a value at a given key is A, and
+/// the client writes B, then immediately reads, it should get B. If something else writes C in
+/// quick succession, then the client may get C. However, a client running in a separate context may
+/// still see A or B
+interface store {
+    /// The set of errors which may be raised by functions in this package
+	variant error {
+		/// The host does not recognize the store identifier requested.
+		no-such-store,
+
+		/// The requesting component does not have access to the specified store
+		/// (which may or may not exist).
+		access-denied,
+
+		/// Some implementation-specific error has occurred (e.g. I/O)
+		other(string)
+	}
+
+    /// A bucket is a collection of key-value pairs. Each key-value pair is stored as a entry in the
+	/// bucket, and the bucket itself acts as a collection of all these entries.
+	///
+	/// It is worth noting that the exact terminology for bucket in key-value stores can very
+	/// depending on the specific implementation. For example:
+	///
+	/// 1. Amazon DynamoDB calls a collection of key-value pairs a table
+	/// 2. Redis has hashes, sets, and sorted sets as different types of collections
+	/// 3. Cassandra calls a collection of key-value pairs a column family
+	/// 4. MongoDB calls a collection of key-value pairs a collection
+	/// 5. Riak calls a collection of key-value pairs a bucket
+	/// 6. Memcached calls a collection of key-value pairs a slab
+	/// 7. Azure Cosmos DB calls a collection of key-value pairs a container
+	///
+	/// In this interface, we use the term `bucket` to refer to a collection of key-value pairs
+    resource bucket {
+        /// Get the store with the specified identifier.
+		///
+		/// `identifier` must refer to a bucket provided by the host.
+		///
+		/// `error::no-such-store` will be raised if the `identifier` is not recognized.
+        open: func(identifier: string) -> result<bucket, error>;
```

Comment:
I think we need to move `open` outside the `resource bucket { ... }`.

Reply:
I see two options of defining `open`:
1. a free function defined outside of the resource as you suggested
2. a `static` function defined inside the resource like `open: static func(identifier: string) -> result<bucket, error>;

I prefer the second one since it's bounded to the scope of the resource type. What do you think?

Reply:
Yeah, that's a good question.  Since the `store` interface itself acts as a scope (which bindings generators will take into account), `open` will already be pretty tightly scoped (in the same way that, e.g., the `handle` function of `outgoing-handler` is), so further nesting `open` inside the resource doesn't additionally help avoid name clashes by further tightening the scope, but it does add a bit of extra syntactic overhead, so I'd default to having `open` just be directly inside the `interface`.  Also, from another perspective, `interface`-scoped functions are like "methods" of the component instance that implements-and-exports that `interface`, and given what `open` does (it looks up a bucket in a keyvalue store), I think it makes sense to have `open` be a "method" of a component implementing a `store`.